---
title: "miRNA discovery - mirdeep2"
author: "Jill Ashey"
date: "2024-11-18"
output: html_document
---

I ran mirdeep2 on R1 reads from DT Mcap 2023 experiment. The output was files with novel and known miRNAs. In this code, I filter the results to identify putative miRNAs!

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(rvest)
```

## Semi raw reads 
NOTE: these miRNAs are from semi-raw reads - I still need to trim adapters from certain reads. 

Load novel miRNAs 
```{r}
novel <- read.csv("../data/Molecular/smRNA/mirdeep2/all_novel.csv", header = T)

length(unique(novel$provisional.id)) # 752 unique genomic ids
length(unique(novel$consensus.mature.sequence)) # 625 unique miRNA sequences
```

Filter so that mirdeep2 score >10, no rfam info, at least 10 reads in mature and star read count, and significant randfold p-value. 
```{r}
novel_filt <- novel %>%
  filter(miRDeep2.score > 10) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(novel_filt$provisional.id))
length(unique(novel_filt$consensus.mature.sequence))

# Save csv 
#write.csv(novel_filt, "../data/Molecular/smRNA/mirdeep2/novel_filt.csv")
```

Load the known miRNAs
```{r}
known <- read.csv("../data/Molecular/smRNA/mirdeep2/all_known.csv")
length(unique(known$tag.id)) # 14 unique genomic ids
length(unique(known$consensus.mature.sequence)) # 14 unique miRNA sequences
```

Because we can see some mapping to known miRNAs in cnidarians, I'm going to filter so that mirdeep2 score >0, no rfam info, at least 10 reads in mature read count, and significant ranfold p-value
```{r}
known_filt <- known %>%
  filter(miRDeep2.score > 0) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(known_filt$tag.id))
length(unique(known_filt$consensus.mature.sequence))

# Save csv 
#write.csv(known_filt, "../data/Molecular/smRNA/mirdeep2/known_filt.csv")
```

## Reads trimmed to 75bp with flexbar 

Load novel miRNAs 
```{r}
novel_75 <- read.csv("../output/Molecular/smRNA/mirdeep2/flexbar_75bp/flexbar_75bp_novel.csv", header = T)

length(unique(novel_75$provisional.id)) # 610 unique genomic ids
length(unique(novel_75$consensus.mature.sequence)) # 507 unique miRNA sequences
```

Filter so that mirdeep2 score >10, no rfam info, at least 10 reads in mature and star read count, and significant randfold p-value. 
```{r}
novel_75_filt <- novel_75 %>%
  filter(miRDeep2.score > 10) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(novel_75_filt$provisional.id))
length(unique(novel_75_filt$consensus.mature.sequence))

# Save csv 
#write.csv(novel_75_filt, "../output/Molecular/smRNA/mirdeep2/flexbar_75bp/flexbar_75bp_novel_filt.csv")
```

Load the known miRNAs
```{r}
known_75 <- read.csv("../output/Molecular/smRNA/mirdeep2/flexbar_75bp/flexbar_75bp_known.csv", header = T)
length(unique(known_75$tag.id)) # 11 unique genomic ids
length(unique(known_75$consensus.mature.sequence)) # 11 unique miRNA sequences
```

Because we can see some mapping to known miRNAs in cnidarians, I'm going to filter so that mirdeep2 score >0, no rfam info, at least 10 reads in mature read count, and significant ranfold p-value
```{r}
known_75_filt <- known_75 %>%
  filter(miRDeep2.score > 0) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(known_75_filt$tag.id))
length(unique(known_75_filt$consensus.mature.sequence))

# Save csv 
#write.csv(known_75_filt, "../output/Molecular/smRNA/mirdeep2/flexbar_75bp/flexbar_75bp_known_filt.csv")
```

## Reads trimmed to 35bp with flexbar 

Load novel miRNAs 
```{r}
novel_35 <- read.csv("../output/Molecular/smRNA/mirdeep2/flexbar_35bp/flexbar_35bp_novel.csv", header = T)

length(unique(novel_35$provisional.id)) # 592 unique genomic ids
length(unique(novel_35$consensus.mature.sequence)) # 497 unique miRNA sequences
```

Filter so that mirdeep2 score >10, no rfam info, at least 10 reads in mature and star read count, and significant randfold p-value. 
```{r}
novel_35_filt <- novel_35 %>%
  filter(miRDeep2.score > 10) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(novel_35_filt$provisional.id))
length(unique(novel_35_filt$consensus.mature.sequence))

# Save csv 
write.csv(novel_35_filt, "../output/Molecular/smRNA/mirdeep2/flexbar_35bp/flexbar_35bp_novel_filt.csv")
```

Load the known miRNAs
```{r}
known_35 <- read.csv("../output/Molecular/smRNA/mirdeep2/flexbar_35bp/flexbar_35bp_known.csv", header = T)
length(unique(known_35$tag.id)) # 10 unique genomic ids
length(unique(known_35$consensus.mature.sequence)) # 10 unique miRNA sequences
```

Because we can see some mapping to known miRNAs in cnidarians, I'm going to filter so that mirdeep2 score >0, no rfam info, at least 10 reads in mature read count, and significant ranfold p-value
```{r}
known_35_filt <- known_35 %>%
  filter(miRDeep2.score > 0) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(known_35_filt$tag.id))
length(unique(known_35_filt$consensus.mature.sequence))

# Save csv 
write.csv(known_35_filt, "../output/Molecular/smRNA/mirdeep2/flexbar_35bp/flexbar_35bp_known_filt.csv")
```

## Trim stringent (cutadapt) first batch novel and known miRNAs 

Load novel miRNAs 
```{r}
novel <- read.csv("../output/Molecular/smRNA/mirdeep2/trim_stringent_first_batch/trim_stringent_first_batch_novel.csv", header = T)

length(unique(novel$provisional.id)) # 659 unique genomic ids
length(unique(novel$consensus.mature.sequence)) # 554 unique miRNA sequences
```

Filter so that mirdeep2 score >10, no rfam info, at least 10 reads in mature and star read count, and significant randfold p-value. 
```{r}
novel_filt <- novel %>%
  filter(miRDeep2.score > 10) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(novel_filt$provisional.id))
length(unique(novel_filt$consensus.mature.sequence))
length(unique(novel_filt$consensus.star.sequence))
length(unique(novel_filt$consensus.precursor.sequence))

# Save csv 
#write.csv(novel_filt, "../output/Molecular/smRNA/mirdeep2/trim_stringent_first_batch/trim_stringent_first_batch_novel_filt.csv")
```
There are 50 novel miRNAs idenfitied but there are 41 unique mature miRNA sequences, which means we either have repeated sequences that are close to one another in the genome OR we have repeated sequences that have unique locations in the genome. I think I also need to look at the pdfs of the miRNAs to determine which ones are putative. 

Look at the repeated sequences 
```{r}
# Count occurrences of each unique mature sequence
sequence_counts <- table(novel_filt$consensus.mature.sequence)

# Find sequences that appear more than once
repeated_sequences <- sequence_counts[sequence_counts > 1]

# Display the repeated sequences and their counts
print(repeated_sequences)

# Find the rows with repeated sequences
repeated_rows <- novel_filt[novel_filt$consensus.mature.sequence %in% names(repeated_sequences), ]

# Display the rows with repeated sequences
print(repeated_rows)
```

`Montipora_capitata_HIv3___Scaffold_14_1001044` and `Montipora_capitata_HIv3___Scaffold_14_1001054` has the same mirdeep2 score and same mature read count, which makes me suspect that there is just one miRNA. Look at the pdfs of each identified miRNA (pdfs located [here](https://github.com/JillAshey/DevelopmentalTimeseries/tree/main/output/Molecular/smRNA/mirdeep2/trim_stringent_first_batch/pdfs_19_02_2025_t_14_06_12)). Looking at both pdfs, they look the exact same...I think these are technically the same. I will remove `Montipora_capitata_HIv3___Scaffold_14_1001054` from the miRNAs. 

`Montipora_capitata_HIv3___Scaffold_9_581562` and `Montipora_capitata_HIv3___Scaffold_10_624116` have the same sequence but are on totally different chromosomes. The mature sequences look the same but the star sequences are slightly different. I feel comfortable saying these are different miRNAs. 







Load the known miRNAs
```{r}
known <- read.csv("../output/Molecular/smRNA/mirdeep2/trim_stringent_first_batch/trim_stringent_first_batch_known.csv", header = T)
length(unique(known$tag.id)) # 15 unique genomic ids
length(unique(known$consensus.mature.sequence)) # 15 unique miRNA sequences
```

Because we can see some mapping to known miRNAs in cnidarians, I'm going to filter so that mirdeep2 score >0, no rfam info, at least 10 reads in mature read count, and significant ranfold p-value
```{r}
known_filt <- known %>%
  filter(miRDeep2.score > 0) %>%
  filter(rfam.alert == "-") %>%
  filter(mature.read.count > 10) %>%
  #filter(star.read.count >10) %>%
  filter(significant.randfold.p.value == "yes") #%>%
  #na.omit()
length(unique(known_filt$tag.id))
length(unique(known_filt$consensus.mature.sequence))

# Save csv 
write.csv(known_filt, "../output/Molecular/smRNA/mirdeep2/trim_stringent_first_batch/trim_stringent_first_batch_known_filt.csv")
```

