---
title: "mRNA differential expression"
author: "Jill Ashey"
date: "2024-10-24"
output: html_document
---

This script provides details on mRNA differential expression across time point in Mcap development.  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

Load packages 
```{r}
library(genefilter)
library(DESeq2)
library(pheatmap)
library(lme4)
library(tidyverse)
library(car)
library(gplots)
library(wesanderson)
library(cowplot)
library(gridExtra)
library(grid)
library("factoextra")
library(vegan)
library(ggcorrplot)
```

Load counts data
```{r}
counts <- read.csv("../data/Molecular/mRNA/Mcap_gene_count_matrix.csv")
counts <- as.data.frame(counts)
rownames(counts) <- counts[,1] #set first column that contains gene names as rownames
counts <- counts[,-1] # remove column w/ gene names 
```

We have 32 columns as expected. Remove extra info from the column names 
```{r}
colnames(counts) <- gsub("^X", "M", sub("_.*", "", colnames(counts)))
```

Remove any genes that have 0 counts across all samples (ie these genes were not expressed)
```{r}
dim(counts) 

counts<-counts %>%
     mutate(Total = rowSums(.[, 1:32]))%>%
    filter(!Total==0)%>%
    dplyr::select(!Total)

dim(counts)
```

Started with 54384 genes, the total number in the annotation. After removing genes with 0s across all samples, 53869 genes are left. 

Filter data with pOverA. There are 32 samples with an n=4 samples per timepoint. Genes will pass filtering if they are present in 4/32=0.125 of the samples because we expect differential expression by timepoint. Also filtering for a minimum gene count of 10, such that ~12% of the samples must have a gene count of >10 in order for the gene to remain in the data set.  

```{r}
filt <- filterfun(pOverA(0.12,10))

#create filter for the counts data
gfilt <- genefilter(counts, filt)

#identify genes to keep by count filter
gkeep <- counts[gfilt,]

#identify gene lists
gn.keep <- rownames(gkeep)

#gene count data filtered in PoverA, P percent of the samples have counts over A
counts_filt <- as.data.frame(counts[which(rownames(counts) %in% gn.keep),])

#How many rows do we have before and after filtering?
nrow(counts) #Before
nrow(counts_filt) #After

# Save filtered counts matrix 
#write.csv(counts_filt, file = "../output/Molecular/mRNA/mRNA_filtered_counts.csv")
```

Before filtering, there were 53869 genes; after filtering, there were 29150 genes. 

Read in metadata 
```{r}
meta <- read.csv("../data/sample_metadata/tube_meta_mcap.csv") %>%
  select(TubeID, Date, hpf, Metric, EmbryosPerTube) %>%
  filter(Metric == "Molecular")

# Filter for sequenced samples
meta_filt <- meta %>%
  filter(TubeID %in% colnames(counts_filt))
```

Reorder columns in filtered counts matrix so that it matches the metadata
```{r}
counts_filt <- counts_filt[, meta_filt$TubeID]

# Sanity check! 
meta_filt$TubeID %in% colnames(counts_filt) #are all of the sample names (rows) in the metadata df in the gene count matrix? Should be TRUE. 
all(rownames(meta_filt$TubeID) == colnames(counts_filt)) #are they the same in the same order? Should be TRUE
```

Set levels of factors 
```{r}
meta_filt$hpf <- factor(meta_filt$hpf, levels = c("1 hpf", "4 hpf", "9 hpf", "14 hpf", "22 hpf", "28 hpf", "48 hpf", "72 hpf"))
```

Construct DESeq data set
```{r}
gdds <- DESeqDataSetFromMatrix(countData = counts_filt, colData = meta_filt, design = ~hpf)
```

Log transform data using VST for visualization purposes, which will deal with low count variability by calculating within-group variation (if blind=FALSE). It doesn't remove design from the data so we can see how all samples look (check if grouping by lifestage, batch effects, etc). To do this, calculate size factors of samples,which is a rough estimate of how many reads each sample contains compared to others. To use VST, size factors should be <4.
```{r}
SF.gdds <- estimateSizeFactors(gdds) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than 4 for us to use vst
print(sizeFactors(SF.gdds)) #View size factors

all(sizeFactors(SF.gdds)) < 4 # should return TRUE
```

All size factors are <4, so we can use vst. 
```{r}
gvst <- vst(gdds, blind=FALSE) #apply a variance stabilizing transformation to minimize effects of small counts and normalize wrt library size
head(assay(gvst), 3) #view transformed gene count data for the first three genes in the dataset.  
```

Plot a heatmap to sample to sample distances  
```{r}
gsampleDists <- dist(t(assay(gvst))) #calculate distance matix
gsampleDistMatrix <- as.matrix(gsampleDists) #distance matrix
rownames(gsampleDistMatrix) <- colnames(gvst) #assign row names
colnames(gsampleDistMatrix) <- NULL #assign col names
#colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) #assign colors

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   pdf(filename, width=width, height=height)
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}

pht<-pheatmap(gsampleDistMatrix, #plot matrix
         clustering_distance_rows=gsampleDists, #cluster rows
         clustering_distance_cols=gsampleDists) #cluster columns
```

Generate a PCA of all genes    
```{r}
pcaData <- plotPCA(gvst, intgroup = c("hpf"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color = hpf, label = name)) + 
  geom_point(size = 3) + 
  geom_text(vjust = 1.5, hjust = 1.5) + # This adds the labels
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
```

PCA separates beautifully!! There are a couple of outliers -- M14 from 4hpf and M61 from 28 hpf. Will need to remove these before proceeding. 

Run test using the LRT to look for effect of lifestage as a whole across our experiment. There is only one main effect, so the reduced model is just the intercept.    
```{r, message = FALSE}
DEG <- DESeq(gdds, test="LRT", reduced=~1)
```

Save all DE results, regardless of significance 
```{r}
DEG_results_all<-as.data.frame(results(DEG))
DEG_results_all$gene <- rownames(DEG_results_all)
rownames(DEG_results_all) <- NULL
dim(DEG_results_all)

# Save as csv 
#write.csv(DEG_results_all, "../output/Molecular/mRNA/mRNA_DE_results_all.csv")
```

Filter by adjusted p value <0.05, this is generated by the LRT test statistic. 
```{r}
DEG_results <- subset(DEG_results_all, padj<0.05)
dim(DEG_results)

# Save as csv 
#write.csv(DEG_results, "../output/Molecular/mRNA/mRNA_DE_results.csv")
```

There are 23977 DEGs out of 29150 total genes expressed (~82%). Most genes are differentially expressed across time, which is expected given that this is a developmental timeseries and lots of molecular changes are happening. 

Conduct VST transformation for visualization.  
```{r}
DEG_results_vst <- gdds[unique(DEG_results$gene)]
dim(DEG_results_vst)

DEG_results_vst <- varianceStabilizingTransformation(DEG_results_vst) 
```

Make a heatmap later 

Generate a PCA of DEGs    
```{r}
pcaData_DEG <- plotPCA(DEG_results_vst, intgroup = c("hpf"), returnData = TRUE)

percentVar_DEG <- round(100 * attr(pcaData_DEG, "percentVar"))
ggplot(pcaData_DEG, aes(PC1, PC2, color = hpf, label = name)) + 
  #geom_point(size = 3) + 
  #geom_jitter(size = 4, width = 1.5, height = 1.5, stroke = 1) +
  geom_jitter(aes(fill = hpf), size = 5, width = 1.5, height = 1.5, shape = 21, color = "black", stroke = 0.5) +
  #geom_text(vjust = 1.5, hjust = 1.5) + # This adds the labels
  xlab(paste0("PC1: ", percentVar_DEG[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar_DEG[2], "% variance")) +
  scale_color_discrete() +
  coord_fixed() +
  theme_linedraw() +
  theme(legend.text = element_text(size = 13),
        plot.background = element_blank(),
        legend.title = element_text(size = 18, face = "bold"),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 15, face = "bold"),
        #axis.title.y = element_text(vjust = -1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5))
```

Make PCA of 1-22hpf
```{r}
hpf_to_keep <- c("1 hpf", "4 hpf", "9 hpf", "14 hpf", "22 hpf")

# Create a logical vector for filtering
keep_samples <- DEG_results_vst$colData$hpf %in% hpf_to_keep

# Subset the DESeqTransform object
filtered_DEG_results_vst <- DEG_results_vst[, keep_samples]

pcaData_DEG <- plotPCA(filtered_DEG_results_vst, intgroup = c("hpf"), returnData = TRUE)

percentVar_DEG <- round(100 * attr(pcaData_DEG, "percentVar"))
ggplot(pcaData_DEG, aes(PC1, PC2, color = hpf, label = name)) + 
  #geom_point(size = 3) + 
  #geom_jitter(size = 4, width = 1.5, height = 1.5, stroke = 1) +
  geom_jitter(aes(fill = hpf), size = 5, width = 1.5, height = 1.5, shape = 21, color = "black", stroke = 0.5) +
  #geom_text(vjust = 1.5, hjust = 1.5) + # This adds the labels
  xlab(paste0("PC1: ", percentVar_DEG[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar_DEG[2], "% variance")) +
  scale_color_discrete() +
  coord_fixed() +
  theme_linedraw() +
  theme(legend.text = element_text(size = 13),
        plot.background = element_blank(),
        legend.title = element_text(size = 18, face = "bold"),
        axis.text = element_text(size = 13),
        axis.title = element_text(size = 15, face = "bold"),
        #axis.title.y = element_text(vjust = -1.5),
        plot.title = element_text(size = 20, face = "italic", hjust = 0.5))
```


Basically the same as above. 

Look at specific contrasts 
```{r}
resultsNames(DEG)
#"hpf_4.hpf_vs_1.hpf"  "hpf_9.hpf_vs_1.hpf"  "hpf_14.hpf_vs_1.hpf"
# "hpf_22.hpf_vs_1.hpf" "hpf_28.hpf_vs_1.hpf" "hpf_48.hpf_vs_1.hpf" "hpf_72.hpf_vs_1.hpf"
```

1 v 4 hpf 
```{r}
results_1v4 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "4 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v4, "../output/Molecular/mRNA/mRNA_DE_results_all_1v4.csv")
```

1 v 9 hpf 
```{r}
results_1v9 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "9 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v9, "../output/Molecular/mRNA/mRNA_DE_results_all_1v9.csv")
```

1 v 14 hpf 
```{r}
results_1v14 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "14 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v14, "../output/Molecular/mRNA/mRNA_DE_results_all_1v14.csv")
```

1 v 22 hpf 
```{r}
results_1v22 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "22 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v22, "../output/Molecular/mRNA/mRNA_DE_results_all_1v22.csv")
```

1 v 28 hpf 
```{r}
results_1v28 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "28 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v28, "../output/Molecular/mRNA/mRNA_DE_results_all_1v28.csv")
```

1 v 48 hpf 
```{r}
results_1v48 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "48 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v48, "../output/Molecular/mRNA/mRNA_DE_results_all_1v48.csv")
```

1 v 72 hpf 
```{r}
results_1v72 <- as.data.frame(results(DEG, contrast = c("hpf", "1 hpf", "72 hpf")))

# Save DE results for all expressed miRNAs
#write.csv(results_1v72, "../output/Molecular/mRNA/mRNA_DE_results_all_1v72.csv")
```


