---
title: "ribofree vs polyA differential expression"
author: "Jill Ashey"
date: "2025-04-29"
output: html_document
---

Looking at differences in differential expression from ribofree and polyA samples 

Load libraries 
```{r}
library(tidyverse)
```

## Counts 

Read in raw counts for ribofree 
```{r}
ribo <- read_csv("../data/Molecular/mRNA/Mcap_gene_count_matrix.csv")
ribo <- as.data.frame(ribo)
rownames(ribo) <- ribo[,1] #set first column that contains gene names as rownames
ribo <- ribo[,-1] # remove column w/ gene names 
colnames(ribo) <- gsub("^X", "M", sub("_.*", "", colnames(ribo))) # remove extra info from col name
colnames(ribo) <- paste0("M", colnames(ribo))
dim(ribo)
```

Read in raw counts for polyA 
```{r}
polyA <- read_csv("../data/Molecular/mRNA_polyA/Mcap_polyA_gene_count_matrix.csv")
polyA <- as.data.frame(polyA)
rownames(polyA) <- polyA[,1] #set first column that contains gene names as rownames
polyA <- polyA[,-1] # remove column w/ gene names 
colnames(polyA) <- gsub(".bam.gtf", "", colnames(polyA))
dim(polyA)
```

I'm not going to do any counts filtering initially. I want to plot the genes over time

Read in metadata
```{r}
meta <- read.csv("../data/sample_metadata/tube_meta_mcap.csv") %>%
  dplyr::select(TubeID, Date, hpf, Metric, EmbryosPerTube) %>%
  filter(Metric == "Molecular")

# Filter for sequenced samples
meta_filt <- meta %>%
  filter(TubeID %in% colnames(polyA))
```

Tidy data and combine ribo and polyA dfs
```{r}
ribo_long <- ribo %>%
  rownames_to_column(var = "GeneID") %>%
  pivot_longer(-GeneID, names_to = "TubeID", values_to = "Counts") %>%
  mutate(Library = "ribo")

polyA_long <- polyA %>%
  rownames_to_column(var = "GeneID") %>%
  pivot_longer(-GeneID, names_to = "TubeID", values_to = "Counts") %>%
  mutate(Library = "polyA")

# Combine
combined_counts <- bind_rows(ribo_long, polyA_long)
```

Add time point metadata
```{r}
meta_filt$TubeID <- as.character(meta_filt$TubeID)

# Join
combined_counts <- combined_counts %>%
  left_join(meta_filt, by = "TubeID")

# Set levels of factors
combined_counts$hpf <- factor(combined_counts$hpf, levels = c("1 hpf", "4 hpf", "9 hpf", "14 hpf", "22 hpf", "28 hpf", "48 hpf", "72 hpf"))
```

Normalize counts data - counts per million
```{r}
library_sizes <- combined_counts %>%
  group_by(Library, TubeID) %>%
  summarize(LibSize = sum(Counts))

# Add CPM
combined_counts <- combined_counts %>%
  left_join(library_sizes, by = c("Library", "TubeID")) %>%
  mutate(CPM = (Counts / LibSize) * 1e6)
```

Plot example gene: 
```{r}
# Subset one gene
gene_of_interest <- "Montipora_capitata_HIv3___TS.g40891.t1b"

gene_plot <- combined_counts %>%
  filter(GeneID == gene_of_interest)

ggplot(gene_plot, aes(x = hpf, y = CPM, color = Library, group = Library)) +
  geom_point() +
  geom_line() +
  labs(title = gene_of_interest,
       x = "Hours post fertilization (hpf)",
       y = "CPM",
       color = "Library") +
  theme_minimal()

```

Calculate mean CPM per gene/timepoint/library
```{r}
mean_cpm <- combined_counts %>%
  group_by(GeneID, hpf, Library) %>%
  summarize(mean_CPM = mean(CPM), .groups = "drop")
```

Pivot table 
```{r}
library_wide <- mean_cpm %>%
  pivot_wider(names_from = Library,
              values_from = mean_CPM,
              values_fill = 0)  # fill missing with 0
```

Compute the ratio (polyA / ribofree for each gene for each timepoint) and plot
```{r}
library_wide <- library_wide %>%
  mutate(Ratio = polyA / (ribo + 1))

# Plot example for one gene
gene_plot <- library_wide %>%
  filter(GeneID == "Montipora_capitata_HIv3___RNAseq.g45549.t1")

ggplot(gene_plot, aes(x = hpf, y = Ratio)) +
  geom_point() +
  geom_line() +
  scale_y_log10() +
  labs(title = "PolyA / rRNA ratio",
       x = "hpf",
       y = "Ratio (polyA / rRNA)",
       subtitle = unique(gene_plot$GeneID)) +
  theme_minimal()
```

Summarize the ratio data based on expected biological outcomes. Here, I am using 4hpf as the early timepoint and 48 hpf as the later timepoint. 
```{r}
library_wide <- library_wide %>%
  mutate(hpf_num = as.numeric(str_remove(as.character(hpf), " hpf")))

ratio_summary <- library_wide %>%
  group_by(GeneID) %>%
  summarize(
    early_ratio = mean(Ratio[hpf_num <= 4], na.rm = TRUE),
    late_ratio = mean(Ratio[hpf_num >= 48], na.rm = TRUE),
    overall_ratio = mean(Ratio, na.rm = TRUE),
    ratio_change = late_ratio - early_ratio,
    .groups = "drop"
  )
ratio_summary <- ratio_summary %>%
  mutate(
    Category = case_when(
      ratio_change < -0.2 & early_ratio > 1 ~ "Drops: Maternal clearance",
      ratio_change > 0.2 & late_ratio > 1 ~ "Rises: Zygotic activation",
      overall_ratio < 0.5 ~ "Stays low: Non-polyA",
      overall_ratio > 1 ~ "Stays high: Stable polyA",
      TRUE ~ "Other/Unclear"
    )
  )
ratio_summary %>%
  count(Category) %>%
  arrange(desc(n))
```

Probably would be best to rlog for expression patterns and CPM/TPM for ratios













## DE 

### 1 v 4 hpf

Read in polyA v ribofree 1v4hpf
```{r}
polyA_de <- read.csv("../output/Molecular/mRNA_polyA/mRNA_DE_results_sig_1v4_wald_polyA.csv")
colnames(polyA_de)[1] <- "gene_id"
colnames(polyA_de)[-1] <- paste0("polyA_", colnames(polyA_de)[-1])

ribofree_de <- read.csv("../output/Molecular/mRNA/mRNA_DE_results_sig_1v4_wald.csv")
colnames(ribofree_de)[1] <- "gene_id"
colnames(ribofree_de)[-1] <- paste0("ribofree_", colnames(ribofree_de)[-1])
```

Find shared and unique genes  
```{r}
# Get the gene IDs from each dataframe
ribofree_genes <- ribofree_de$gene_id
polya_genes <- polyA_de$gene_id

# Shared genes
shared_genes <- intersect(ribofree_genes, polya_genes)

# Unique to ribofree
unique_ribofree <- setdiff(ribofree_genes, polya_genes)

# Unique to polyA
unique_polya <- setdiff(polya_genes, ribofree_genes)

# New df with shared genes 
ribofree_shared <- ribofree_de[ribofree_de$gene_id %in% shared_genes, ]
polya_shared <- polyA_de[polyA_de$gene_id %in% shared_genes, ]

# New df with unique genes 
ribofree_unique <- ribofree_de[ribofree_de$gene_id %in% unique_ribofree, ]
polya_unique <- polyA_de[polyA_de$gene_id %in% unique_polya, ]
```

Merge shared dfs 
```{r}
# Merge shared genes dataframes
shared_de <- merge(polya_shared, ribofree_shared, by = "gene_id")

# Classify genes based on expression changes
shared_de <- shared_de %>%
  mutate(regulation_category = case_when(
    polyA_regulation == "Up at 4 hpf" & ribofree_regulation == "Up at 4 hpf" ~ "Transcriptional Up",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Transcriptional Down",
    polyA_regulation == "Up at 4 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Deadenylation",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 4 hpf" ~ "Readenylation",
    polyA_regulation == "Up at 4 hpf" & ribofree_regulation == "NS" ~ "PolyA Increase, Transcription Unchanged",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "NS" ~ "PolyA Decrease, Transcription Unchanged",
    ribofree_regulation == "Up at 4 hpf" & polyA_regulation == "NS" ~ "Ribofree Increase, PolyA Unchanged",
    ribofree_regulation == "Up at 1 hpf" & polyA_regulation == "NS" ~ "Ribofree Decrease, PolyA Unchanged",
    TRUE ~ "Complex/Other"
  ))

# Summarize regulation categories
category_summary <- shared_de %>%
  group_by(regulation_category) %>%
  summarize(count = n())
```

Read in annot and merge with shared and unique dfs 
```{r}
annot <- read.delim("~/Desktop/GFFs/mcap/V3/Montipora_capitata_HIv3.genes.EggNog_results.txt")
colnames(annot)[1] <- "gene_id"

# Shared 
merge_annot <- merge %>%
  inner_join(annot, by = "gene_id")

# Unique to ribofree
merge_ribofree <- ribofree_unique %>%
    inner_join(annot, by = "gene_id")
```

### 1 v 9 hpf

Read in polyA v ribofree 1v9hpf
```{r}
polyA_de <- read.csv("../output/Molecular/mRNA_polyA/mRNA_DE_results_sig_1v9_wald_polyA.csv")
colnames(polyA_de)[1] <- "gene_id"
colnames(polyA_de)[-1] <- paste0("polyA_", colnames(polyA_de)[-1])

ribofree_de <- read.csv("../output/Molecular/mRNA/mRNA_DE_results_sig_1v9_wald.csv")
colnames(ribofree_de)[1] <- "gene_id"
colnames(ribofree_de)[-1] <- paste0("ribofree_", colnames(ribofree_de)[-1])
```

Find shared and unique genes  
```{r}
# Get the gene IDs from each dataframe
ribofree_genes <- ribofree_de$gene_id
polya_genes <- polyA_de$gene_id

# Shared genes
shared_genes <- intersect(ribofree_genes, polya_genes)

# Unique to ribofree
unique_ribofree <- setdiff(ribofree_genes, polya_genes)

# Unique to polyA
unique_polya <- setdiff(polya_genes, ribofree_genes)

# New df with shared genes 
ribofree_shared <- ribofree_de[ribofree_de$gene_id %in% shared_genes, ]
polya_shared <- polyA_de[polyA_de$gene_id %in% shared_genes, ]

# New df with unique genes 
ribofree_unique <- ribofree_de[ribofree_de$gene_id %in% unique_ribofree, ]
polya_unique <- polyA_de[polyA_de$gene_id %in% unique_polya, ]
```

Merge shared dfs 
```{r}
# Merge shared genes dataframes
shared_de <- merge(polya_shared, ribofree_shared, by = "gene_id")

# Classify genes based on expression changes
shared_de <- shared_de %>%
  mutate(regulation_category = case_when(
    polyA_regulation == "Up at 9 hpf" & ribofree_regulation == "Up at 9 hpf" ~ "Transcriptional Up",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Transcriptional Down",
    polyA_regulation == "Up at 9 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Deadenylation",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 9 hpf" ~ "Readenylation",
    polyA_regulation == "Up at 9 hpf" & ribofree_regulation == "NS" ~ "PolyA Increase, Transcription Unchanged",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "NS" ~ "PolyA Decrease, Transcription Unchanged",
    ribofree_regulation == "Up at 9 hpf" & polyA_regulation == "NS" ~ "Ribofree Increase, PolyA Unchanged",
    ribofree_regulation == "Up at 1 hpf" & polyA_regulation == "NS" ~ "Ribofree Decrease, PolyA Unchanged",
    TRUE ~ "Complex/Other"
  ))

# Summarize regulation categories
category_summary <- shared_de %>%
  group_by(regulation_category) %>%
  summarize(count = n())
```

### 1 v 14 hpf

Read in polyA v ribofree 1v14hpf
```{r}
polyA_de <- read.csv("../output/Molecular/mRNA_polyA/mRNA_DE_results_sig_1v14_wald_polyA.csv")
colnames(polyA_de)[1] <- "gene_id"
colnames(polyA_de)[-1] <- paste0("polyA_", colnames(polyA_de)[-1])

ribofree_de <- read.csv("../output/Molecular/mRNA/mRNA_DE_results_sig_1v14_wald.csv")
colnames(ribofree_de)[1] <- "gene_id"
colnames(ribofree_de)[-1] <- paste0("ribofree_", colnames(ribofree_de)[-1])
```

Find shared and unique genes  
```{r}
# Get the gene IDs from each dataframe
ribofree_genes <- ribofree_de$gene_id
polya_genes <- polyA_de$gene_id

# Shared genes
shared_genes <- intersect(ribofree_genes, polya_genes)

# Unique to ribofree
unique_ribofree <- setdiff(ribofree_genes, polya_genes)

# Unique to polyA
unique_polya <- setdiff(polya_genes, ribofree_genes)

# New df with shared genes 
ribofree_shared <- ribofree_de[ribofree_de$gene_id %in% shared_genes, ]
polya_shared <- polyA_de[polyA_de$gene_id %in% shared_genes, ]

# New df with unique genes 
ribofree_unique <- ribofree_de[ribofree_de$gene_id %in% unique_ribofree, ]
polya_unique <- polyA_de[polyA_de$gene_id %in% unique_polya, ]
```

Merge shared dfs 
```{r}
# Merge shared genes dataframes
shared_de <- merge(polya_shared, ribofree_shared, by = "gene_id")

# Classify genes based on expression changes
shared_de <- shared_de %>%
  mutate(regulation_category = case_when(
    polyA_regulation == "Up at 14 hpf" & ribofree_regulation == "Up at 14 hpf" ~ "Transcriptional Up",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Transcriptional Down",
    polyA_regulation == "Up at 14 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Deadenylation",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 14 hpf" ~ "Readenylation",
    polyA_regulation == "Up at 14 hpf" & ribofree_regulation == "NS" ~ "PolyA Increase, Transcription Unchanged",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "NS" ~ "PolyA Decrease, Transcription Unchanged",
    ribofree_regulation == "Up at 14 hpf" & polyA_regulation == "NS" ~ "Ribofree Increase, PolyA Unchanged",
    ribofree_regulation == "Up at 1 hpf" & polyA_regulation == "NS" ~ "Ribofree Decrease, PolyA Unchanged",
    TRUE ~ "Complex/Other"
  ))

# Summarize regulation categories
category_summary <- shared_de %>%
  group_by(regulation_category) %>%
  summarize(count = n())
```

### 1 v 22 hpf

Read in polyA v ribofree 1v22hpf
```{r}
polyA_de <- read.csv("../output/Molecular/mRNA_polyA/mRNA_DE_results_sig_1v22_wald_polyA.csv")
colnames(polyA_de)[1] <- "gene_id"
colnames(polyA_de)[-1] <- paste0("polyA_", colnames(polyA_de)[-1])

ribofree_de <- read.csv("../output/Molecular/mRNA/mRNA_DE_results_sig_1v22_wald.csv")
colnames(ribofree_de)[1] <- "gene_id"
colnames(ribofree_de)[-1] <- paste0("ribofree_", colnames(ribofree_de)[-1])
```

Find shared and unique genes  
```{r}
# Get the gene IDs from each dataframe
ribofree_genes <- ribofree_de$gene_id
polya_genes <- polyA_de$gene_id

# Shared genes
shared_genes <- intersect(ribofree_genes, polya_genes)

# Unique to ribofree
unique_ribofree <- setdiff(ribofree_genes, polya_genes)

# Unique to polyA
unique_polya <- setdiff(polya_genes, ribofree_genes)

# New df with shared genes 
ribofree_shared <- ribofree_de[ribofree_de$gene_id %in% shared_genes, ]
polya_shared <- polyA_de[polyA_de$gene_id %in% shared_genes, ]

# New df with unique genes 
ribofree_unique <- ribofree_de[ribofree_de$gene_id %in% unique_ribofree, ]
polya_unique <- polyA_de[polyA_de$gene_id %in% unique_polya, ]
```

Merge shared dfs 
```{r}
# Merge shared genes dataframes
shared_de <- merge(polya_shared, ribofree_shared, by = "gene_id")

# Classify genes based on expression changes
shared_de <- shared_de %>%
  mutate(regulation_category = case_when(
    polyA_regulation == "Up at 22 hpf" & ribofree_regulation == "Up at 22 hpf" ~ "Transcriptional Up",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Transcriptional Down",
    polyA_regulation == "Up at 22 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Deadenylation",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 22 hpf" ~ "Readenylation",
    polyA_regulation == "Up at 22 hpf" & ribofree_regulation == "NS" ~ "PolyA Increase, Transcription Unchanged",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "NS" ~ "PolyA Decrease, Transcription Unchanged",
    ribofree_regulation == "Up at 22 hpf" & polyA_regulation == "NS" ~ "Ribofree Increase, PolyA Unchanged",
    ribofree_regulation == "Up at 1 hpf" & polyA_regulation == "NS" ~ "Ribofree Decrease, PolyA Unchanged",
    TRUE ~ "Complex/Other"
  ))

# Summarize regulation categories
category_summary <- shared_de %>%
  group_by(regulation_category) %>%
  summarize(count = n())
```

### 1 v 72 hpf

Read in polyA v ribofree 1v72hpf
```{r}
polyA_de <- read.csv("../output/Molecular/mRNA_polyA/mRNA_DE_results_sig_1v72_wald_polyA.csv")
colnames(polyA_de)[1] <- "gene_id"
colnames(polyA_de)[-1] <- paste0("polyA_", colnames(polyA_de)[-1])

ribofree_de <- read.csv("../output/Molecular/mRNA/mRNA_DE_results_sig_1v72_wald.csv")
colnames(ribofree_de)[1] <- "gene_id"
colnames(ribofree_de)[-1] <- paste0("ribofree_", colnames(ribofree_de)[-1])
```

Find shared and unique genes  
```{r}
# Get the gene IDs from each dataframe
ribofree_genes <- ribofree_de$gene_id
polya_genes <- polyA_de$gene_id

# Shared genes
shared_genes <- intersect(ribofree_genes, polya_genes)

# Unique to ribofree
unique_ribofree <- setdiff(ribofree_genes, polya_genes)

# Unique to polyA
unique_polya <- setdiff(polya_genes, ribofree_genes)

# New df with shared genes 
ribofree_shared <- ribofree_de[ribofree_de$gene_id %in% shared_genes, ]
polya_shared <- polyA_de[polyA_de$gene_id %in% shared_genes, ]

# New df with unique genes 
ribofree_unique <- ribofree_de[ribofree_de$gene_id %in% unique_ribofree, ]
polya_unique <- polyA_de[polyA_de$gene_id %in% unique_polya, ]
```

Merge shared dfs 
```{r}
# Merge shared genes dataframes
shared_de <- merge(polya_shared, ribofree_shared, by = "gene_id")

# Classify genes based on expression changes
shared_de <- shared_de %>%
  mutate(regulation_category = case_when(
    polyA_regulation == "Up at 72 hpf" & ribofree_regulation == "Up at 72 hpf" ~ "Transcriptional Up",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Transcriptional Down",
    polyA_regulation == "Up at 72 hpf" & ribofree_regulation == "Up at 1 hpf" ~ "Deadenylation",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "Up at 72 hpf" ~ "Readenylation",
    polyA_regulation == "Up at 72 hpf" & ribofree_regulation == "NS" ~ "PolyA Increase, Transcription Unchanged",
    polyA_regulation == "Up at 1 hpf" & ribofree_regulation == "NS" ~ "PolyA Decrease, Transcription Unchanged",
    ribofree_regulation == "Up at 72 hpf" & polyA_regulation == "NS" ~ "Ribofree Increase, PolyA Unchanged",
    ribofree_regulation == "Up at 1 hpf" & polyA_regulation == "NS" ~ "Ribofree Decrease, PolyA Unchanged",
    TRUE ~ "Complex/Other"
  ))

# Summarize regulation categories
category_summary <- shared_de %>%
  group_by(regulation_category) %>%
  summarize(count = n())
```



